app/src/main/java/org/mozilla/fenix/components/toolbar/CFRPresenter.kt

```diff
@@ -0,0 +1,130 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.components.toolbar
+
+import android.content.Context
+import androidx.annotation.VisibleForTesting
+import androidx.compose.foundation.clickable
+import androidx.compose.material.Text
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.text.style.TextDecoration
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.flow.mapNotNull
+import kotlinx.coroutines.flow.transformWhile
+import mozilla.components.browser.state.selector.findCustomTabOrSelectedTab
+import mozilla.components.browser.state.store.BrowserStore
+import mozilla.components.browser.toolbar.BrowserToolbar
+import mozilla.components.lib.state.ext.flowScoped
+import mozilla.components.service.glean.private.NoExtras
+import org.mozilla.fenix.GleanMetrics.TrackingProtection
+import org.mozilla.fenix.R
+import org.mozilla.fenix.compose.cfr.CFRPopup
+import org.mozilla.fenix.compose.cfr.CFRPopupProperties
+import org.mozilla.fenix.ext.components
+import org.mozilla.fenix.settings.SupportUtils
+import org.mozilla.fenix.settings.SupportUtils.SumoTopic.TOTAL_COOKIE_PROTECTION
+import org.mozilla.fenix.theme.FirefoxTheme
+import org.mozilla.fenix.utils.Settings
+
+/**
+ * Delegate for handling all the business logic for showing CFRs in the toolbar.
+ *
+ * @param context [Context] used for various Android interactions.
+ * @param store [BrowserStore] which will be observed for tabs updates
+ * @param settings [Settings] allowing to read and write persistent user settings
+ * @param toolbar [BrowserToolbar] that will serve as anchor for the CFRs
+ * @param sessionId [String] optional custom tab id.
+ */
+class CFRPresenter(
+    private val context: Context,
+    private val store: BrowserStore,
+    private val settings: Settings,
+    private val toolbar: BrowserToolbar,
+    private val sessionId: String? = null
+) {
+    @VisibleForTesting
+    internal var tcpCfrScope: CoroutineScope? = null
+    @VisibleForTesting
+    internal var tcpCfr: CFRPopup? = null
+
+    /**
+     * Start observing [store] for updates which may trigger showing a CFR.
+     */
+    @Suppress("MagicNumber")
+    fun start() {
+        if (settings.shouldShowTotalCookieProtectionCFR) {
+            tcpCfrScope = store.flowScoped { flow ->
+                flow
+                    .mapNotNull { it.findCustomTabOrSelectedTab(sessionId)?.content?.progress }
+                    // The "transformWhile" below ensures that the 100% progress is only collected once.
```

Comment:
Wouldn't `takeWhile` be more suitable? We want to receive items until it fits out condition (being == 100), and `takeWhile` does  just that. Something like:

```
                    .takeWhile { progress ->
                        val isLoading = progress != 100
                        if (!isLoading) {
                            showTcpCfr()
                        }
                        isLoading
                    }
                    .collect()
```

Reply:
The above code would indeed work but I don't like executing the work in a filtering operation while collect which is a terminal operator is empty.

Reply:
Fair enough! Do we really need `transformWhile` if we cancel the scope anyway? Wouldn't something like this work? It does pass the tests.
```
                    .filter { it == 100 }
                    .collect{
                        tcpCfrScope?.cancel()
                        showTcpCfr()
                    }
```

Reply:
Progress 100 is emitted multiple times and I saw in testing that this would result in multiple CFRs being shown.
Looks like there is time for the flow to emit a new item.

With the following code
```kotlin
tcpCfrScope = browserStore.flowScoped { flow ->
                flow
                    .mapNotNull { it.findCustomTabOrSelectedTab(sessionId)?.content?.progress }
                    // The "transformWhile" below ensures that the 100% progress is only collected once.
                    // .transformWhile { progress ->
                    //     emit(progress)
                    //     progress != 100
                    // }
                    .filter { it == 100 }
                    .collect {
                        tcpCfrScope?.cancel()
                        println("Mugurel: Showing the CFR")
                        showTcpCfr()
                    }
            }
```

I would always see two of those logs printed, eg:
> 2022-08-12 09:20:45.425 org.mozilla.fenix.debug I/System.out: Mugurel: Showing the CFR
2022-08-12 09:20:45.426 org.mozilla.fenix.debug I/System.out: Mugurel: Showing the CFR

Reply:
thanks for clarifying. I though tests should also catch that behavior, not just manual testing, especially this one

`GIVEN the TCP CFR should be shown WHEN the current tab is fully loaded THEN the TCP CFR is only shown once`